File: server/src/agents/documentComparisonAgent.js
import { BaseAgent } from './baseAgent.js';
import { generateWithContinuation } from '../utils/completionGuard.js';

const OLLAMA_MODEL = process.env.OLLAMA_MODEL || 'llama3.2:3b';
const MAX_DOC_CHARS = 4000; // per document, to keep both docs + prompt within a small model's context window

const SYSTEM_PROMPT = `You are Nova AI's Document Comparison Agent. You are given two documents and must compare them directly — never substitute outside knowledge for either document's actual content.

Reply with this exact markdown structure:
## Agreements
Bullet list of points both documents state consistently.
## Differences
Bullet list of points where the documents disagree or one states something the other doesn't, naming which document says what.
## Summary
One paragraph: the overall relationship between the two documents (e.g. one supersedes the other, they cover different scope, they conflict on a key point).`;

// Document-grounded comparison (the "compare these documents" request
// documentAnalysisAgent.js doesn't handle — that agent analyzes ONE
// document's content, this one diffs TWO). Bypasses web search for the
// same reason: Uploaded Document > Web Search whenever a file is attached.
export class DocumentComparisonAgent extends BaseAgent {
  constructor() {
    super('DocumentComparisonAgent');
  }

  async run(query, context = {}) {
    const { documents = [] } = context; // [{ fileName, text }, ...]
    if (documents.length < 2) {
      return { success: true, output: { answer: 'Comparison needs at least two uploaded documents in this chat — only one was found.' } };
    }

    const [docA, docB] = documents.slice(-2);
    const truncate = (text) => text.length > MAX_DOC_CHARS ? `${text.slice(0, MAX_DOC_CHARS)}\n[...truncated]` : text;

    const prompt = `Document A (${docA.fileName}):\n${truncate(docA.text)}\n\nDocument B (${docB.fileName}):\n${truncate(docB.text)}\n\nRequest: ${query}`;

    try {
      const answer = await generateWithContinuation(
        [
          { role: 'system', content: SYSTEM_PROMPT },
          { role: 'user', content: prompt }
        ],
        { model: OLLAMA_MODEL, numPredict: 1500, temperature: 0.3 }
      );
      return { success: true, output: { answer, comparedFiles: [docA.fileName, docB.fileName] } };
    } catch (err) {
      return { success: false, output: { answer: '' }, error: err.message };
    }
  }
}

File: server/app/agents/document_comparison_agent.py
from ..core.config import OLLAMA_MODEL
from ..utils.completion_guard import generate_with_continuation
from .base_agent import BaseAgent

MAX_DOC_CHARS = 4000

SYSTEM_PROMPT = """You are Nova AI's Document Comparison Agent. You are given two documents and must compare them directly — never substitute outside knowledge for either document's actual content.

Reply with this exact markdown structure:
## Agreements
Bullet list of points both documents state consistently.
## Differences
Bullet list of points where the documents disagree or one states something the other doesn't, naming which document says what.
## Summary
One paragraph: the overall relationship between the two documents (e.g. one supersedes the other, they cover different scope, they conflict on a key point)."""


def _truncate(text: str) -> str:
    return f"{text[:MAX_DOC_CHARS]}\n[...truncated]" if len(text) > MAX_DOC_CHARS else text


class DocumentComparisonAgent(BaseAgent):
    """Document-grounded comparison (the "compare these documents" request).
    Bypasses web search: Uploaded Document > Web Search whenever a file is attached."""

    def __init__(self):
        super().__init__("DocumentComparisonAgent")

    async def run(self, query: str, context: dict | None = None) -> dict:
        context = context or {}
        documents = context.get("documents", [])  # [{fileName, text}, ...]
        if len(documents) < 2:
            return {"success": True, "output": {"answer": "Comparison needs at least two uploaded documents in this chat — only one was found."}}

        doc_a, doc_b = documents[-2:]
        prompt = (
            f"Document A ({doc_a['fileName']}):\n{_truncate(doc_a['text'])}\n\n"
            f"Document B ({doc_b['fileName']}):\n{_truncate(doc_b['text'])}\n\nRequest: {query}"
        )

        try:
            answer = await generate_with_continuation(
                [
                    {"role": "system", "content": SYSTEM_PROMPT},
                    {"role": "user", "content": prompt},
                ],
                model=OLLAMA_MODEL,
                num_predict=1500,
                temperature=0.3,
            )
            return {"success": True, "output": {"answer": answer, "comparedFiles": [doc_a["fileName"], doc_b["fileName"]]}}
        except Exception as err:
            return {"success": False, "output": {"answer": ""}, "error": str(err)}
